src/modal/modal-stack.ts

```diff
@@ -20,75 +22,110 @@ import {NgbActiveModal, NgbModalRef} from './modal-ref';
 
   constructor(
-      private _applicationRef: ApplicationRef, private _injector: Injector,
+      @Inject(DOCUMENT) document, private _applicationRef: ApplicationRef, private _injector: Injector,
       private _componentFactoryResolver: ComponentFactoryResolver) {
-    this._backdropFactory = _componentFactoryResolver.resolveComponentFactory(NgbModalBackdrop);
-    this._windowFactory = _componentFactoryResolver.resolveComponentFactory(NgbModalWindow);
+    this._document = document;
   }
 
-  open(moduleCFR: ComponentFactoryResolver, contentInjector: Injector, content: any, options): NgbModalRef {
-    const containerSelector = options.container || 'body';
-    const containerEl = document.querySelector(containerSelector);
+  open(content: any, options): NgbModalRef {
+    this._buildFactories();
+    const containerEl =
+        isDefined(options.container) ? this._document.querySelector(options.container) : this._document.body;
 
     if (!containerEl) {
-      throw new Error(`The specified modal container "${containerSelector}" was not found in the DOM.`);
+      throw new Error(`The specified modal container "${options.container}" was not found in the DOM.`);
     }
 
     const activeModal = new NgbActiveModal();
-    const contentRef = this._getContentRef(moduleCFR, options.injector || contentInjector, content, activeModal);
+    const contentRef = this._getContentRef(options.injector || this._injector, content, activeModal);
 
-    let windowCmptRef: ComponentRef<NgbModalWindow>;
-    let backdropCmptRef: ComponentRef<NgbModalBackdrop>;
-    let ngbModalRef: NgbModalRef;
-
-
-    if (options.backdrop !== false) {
-      backdropCmptRef = this._backdropFactory.create(this._injector);
-      this._applicationRef.attachView(backdropCmptRef.hostView);
-      containerEl.appendChild(backdropCmptRef.location.nativeElement);
-    }
-    windowCmptRef = this._windowFactory.create(this._injector, contentRef.nodes);
-    this._applicationRef.attachView(windowCmptRef.hostView);
-    containerEl.appendChild(windowCmptRef.location.nativeElement);
-
-    ngbModalRef = new NgbModalRef(windowCmptRef, contentRef, backdropCmptRef, options.beforeDismiss);
+    let backdropCmptRef: ComponentRef<NgbModalBackdrop> =
+        options.backdrop !== false ? this._attachBackdrop(containerEl) : null;
+    let windowCmptRef: ComponentRef<NgbModalWindow> = this._attachWindowComponent(containerEl, contentRef);
+    let ngbModalRef: NgbModalRef = new NgbModalRef(windowCmptRef, contentRef, backdropCmptRef, options.beforeDismiss);
 
     activeModal.close = (result: any) => { ngbModalRef.close(result); };
     activeModal.dismiss = (reason: any) => { ngbModalRef.dismiss(reason); };
 
     this._applyWindowOptions(windowCmptRef.instance, options);
-
     return ngbModalRef;
   }
 
+  // Cant build factories on constructor when using lazy loading
+  private _buildFactories() {
+    this._buildBackdropFactory();
+    this._buildWindowFactory();
```

Comment:
I don't think that caching factories here, in the code, adds any performance benefits. Remove those methods and replace with a direct call to `this._componentFactoryResolver.resolveComponentFactory(NgbModalBackdrop)`